hi-ml/src/health_ml/utils/bag_utils.py

```diff
@@ -0,0 +1,221 @@
+                 max_bag_size: int = 0,
+                 generator: Optional[torch.Generator] = None) -> None:
+        """
+        :param bag_ids: The bag IDs for each sample, of the same length as the dataset.
+        :param shuffle_bags: Whether the bags should be iterated in random order.
+        :param shuffle_samples: Whether the instances in each bag should be shuffled.
+        :param max_bag_size: Upper bound on number of instances in each loaded bag. If 0 (default),
+        will return all samples in each bag. If > 0 and `shuffle_samples=True`, bags larger than
+        `max_bag_size` will yield random subsets of instances. Note that setting `max_bag_size > 0`
+        with `shuffle_samples=False` will return fixed subsets and may completely exclude some
+        samples from the iteration.
+        :param generator: The PRNG to use for shuffling. By default, creates one with a random seed.
+        """
+        self.unique_bag_ids, self.bag_indices = np.unique(bag_ids, return_inverse=True)
+        self.shuffle_bags = shuffle_bags
+        self.shuffle_samples = shuffle_samples
+        self.max_bag_size = max_bag_size
+        self.generator = generator
+
+    def __iter__(self) -> Iterator[List[int]]:
+        generator = self.generator or _create_generator()
+        n_bags = len(self.unique_bag_ids)
+        bag_sequence = torch.randperm(n_bags, generator=generator).tolist() \
+            if self.shuffle_bags else range(n_bags)
+        for bag_idx in bag_sequence:
+            yield self.get_bag(bag_idx, generator)
+
+    def get_bag(self, bag_index: int, generator: Optional[torch.Generator] = None) \
+            -> List[int]:
+        bag, = np.where(self.bag_indices == bag_index)
+        if self.shuffle_samples:
+            if generator is None:
+                generator = self.generator or _create_generator()
+            perm = torch.randperm(len(bag), generator=generator)
+            bag = np.atleast_1d(bag[perm])  # pytorch squeezes singleton tensors
+        if self.max_bag_size > 0:
+            bag = bag[:self.max_bag_size]
+        return bag.tolist()
+
+    def __len__(self) -> int:
+        return len(self.unique_bag_ids)
+
+    def __getstate__(self) -> Dict:
+        # torch.Generator is not pickleable, so we need to serialise only its internal state.
+        # This is useful e.g. for caching and restoring a BagDataset.
+        d = self.__dict__.copy()
+        if self.generator is not None:
+            generator_state = self.generator.get_state() if self.generator is not None else None
+            d['generator'] = generator_state
+        return d
+
+    def __setstate__(self, d: Dict) -> None:
+        # Same here for restoring the torch.Generator state
+        self.__dict__ = d
+        generator = None
+        if d['generator'] is not None:
+            generator = torch.Generator()
+            generator.set_state(d['generator'])
+        self.generator = generator
```

Comment:
Does the deepmil code ever use a generator argument for the BAgDataset? If not, maybe kill all this get_state/set_state code, and simplify?